packages/react-widgets/src/widgets/TimeSeriesWidget.js

```diff
@@ -0,0 +1,308 @@
+
+  useEffect(() => {
+    if (isSourceReady) {
+      setIsLoading(true);
+      const _filters = getApplicableFilters(filters, id);
+
+      getTimeSeries({
+        filters: _filters,
+        dataSource,
+        column,
+        stepSize: selectedStepSize,
+        operationColumn,
+        operation
+      })
+        .then((data) => {
+          if (data) {
+            setIsLoading(false);
+            setTimeSeriesData(data);
+          }
+        })
+        .catch((error) => {
+          setIsLoading(false);
+          if (onError) onError(error);
+        });
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [
+    id,
+    // filters is always a new object, then we have to stringify it
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    JSON.stringify(filters),
+    dataSource,
+    column,
+    selectedStepSize,
+    isSourceReady,
+    setIsLoading,
+    operationColumn,
+    operation,
+    onError
+  ]);
+
+  const handleTimeframeUpdate = useCallback(
+    (timeframe) => {
+      dispatch(
+        addFilter({
+          id: dataSource,
+          column,
+          type: FilterTypes.BETWEEN,
+          values: [timeframe],
+          owner: id
+        })
+      );
+
+      if (onTimeframeUpdate) onTimeframeUpdate();
+    },
+    [column, dataSource, dispatch, id, onTimeframeUpdate]
+  );
+
+  const handleTimelineUpdate = useCallback(
+    (timelinePosition) => {
+      const { name: moment } = timeSeriesData[timelinePosition];
+      handleTimeframeUpdate([
+        moment,
+        moment + BUCKET_SIZE_RANGE_MAPPING[selectedStepSize] * 1000
+      ]);
+
+      if (onTimelineUpdate) onTimelineUpdate();
+    },
+    [handleTimeframeUpdate, onTimelineUpdate, selectedStepSize, timeSeriesData]
+  );
+
+  const handleStop = useCallback(() => {
+    dispatch(
+      removeFilter({
+        id: dataSource,
+        column
+      })
+    );
+
+    if (onStop) onStop();
+  }, [column, dataSource, dispatch, onStop]);
+
+  const [anchorEl, setAnchorEl] = useState(null);
+
+  const handleOpenBucketSizeMenu = (e) => {
+    if (e?.currentTarget) {
+      setAnchorEl(e.currentTarget);
+    }
+  };
+
+  const handleCloseBucketSizeMenu = () => {
+    setAnchorEl(null);
+  };
+
+  const handleStepSizeUpdate = (stepSize) => {
+    setSelectedStepSize(stepSize);
+    handleCloseBucketSizeMenu();
+  };
+
+  return (
+    <>
+      <WrapperWidgetUI
+        title={title}
+        isLoading={isLoading}
+        {...wrapperProps}
+        actions={[
+          ...(wrapperProps?.actions || []),
+          ...(stepSizeOptions?.length
+            ? [
+                {
+                  id: 'a0',
+                  name: 'Bucket size',
+                  icon: <StepSizeIcon />,
+                  action: handleOpenBucketSizeMenu
+                }
+              ]
+            : [])
+        ]}
+      >
+        {!timeSeriesData.length ? (
+          <NoDataAlert {...noDataAlertProps} />
```

Comment:
NoDataAlert shouldn't be used in UI, I think Widget is the place. But I don't want to modify CategoryWidgetUI in this PR, that's why I just generalise that part of the code in a component, and then exported it. Also, I think it's useful to export NoDataAlert because it can be used for users.

Reply:
In that case, we should move NoDataAlert to react-widgets module y modify the CategoryWidgetUI and CategoryWidget to keep the coherence of the modules

Reply:
Should I open a PR with this? I would bring this feature to every widget.

Reply:
Depends on https://github.com/CartoDB/carto-react/pull/168#discussion_r719470231

Reply:
Yes please, open that PR so it lands **before** this one, and then adapt this one to it

Reply:
Done